src/Catalyst.Core.Modules.Dfs/DfsService.cs

```diff
@@ -0,0 +1,530 @@
+            BootstrapApi = bootstrapApi;
+            ConfigApi = configApi;
+            DagApi = dagApi;
+            DhtApi = dhtApi;
+            DhtService = dhtService;
+            UnixFsApi = unixFsApi;
+            KeyApi = keyApi;
+            NameApi = nameApi;
+            ObjectApi = objectApi;
+            PinApi = pinApi;
+            PingService = pingService;
+            PubSubApi = pubSubApi;
+            PubSubService = pubSubService;
+            StatsApi = statsApi;
+            SwarmApi = swarmApi;
+            SwarmService = swarmService;
+            Options = dfsOptions;
+            _hashProvider = hashProvider;
+            _dfsState = dfsState;
+            DnsApi = dnsApi;
+            MigrationManager = migrationManager;
+            LocalPeer = localPeer;
+
+            InitAsync().Wait();
+        }
+
+        internal virtual AddFileOptions AddFileOptions()
+        {
+            return new AddFileOptions
+            {
+                Hash = _hashProvider.HashingAlgorithm.Name,
+                RawLeaves = true
+            };
+        }
+
+        /// <summary>
+        ///     The configuration options.
+        /// </summary>
+        public DfsOptions Options { get; set; }
+
+        //private IKeyStoreService _keyStoreService;
+
+        private async Task InitAsync()
+        {
+            //Swarm
+            Log.Debug("Building swarm service");
+
+            SwarmService.LocalPeer = LocalPeer;
+            SwarmService.LocalPeerKey =
+                Key.CreatePrivateKey(await KeyApi.GetPrivateKeyAsync("self").ConfigureAwait(false));
+            SwarmService.NetworkProtector = Options.Swarm.PrivateNetworkKey == null
+                ? null
+                : new Psk1Protector
+                {
+                    Key = Options.Swarm.PrivateNetworkKey
+                };
+
+            if (Options.Swarm.PrivateNetworkKey != null)
+            {
+                Log.Debug($"Private network {Options.Swarm.PrivateNetworkKey.Fingerprint().ToHexString()}");
+            }
+
+            Log.Debug("Building PubSub service");
+
+            //PubSubService.LocalPeer = LocalPeer;
+            Log.Debug("Built PubSub service");
+        }
+
+        /// <summary>
+        ///     Starts the network services.
+        /// </summary>
+        /// <returns>
+        ///     A task that represents the asynchronous operation.
+        /// </returns>
+        /// <remarks>
+        ///     Starts the various IPFS and Lib.P2P network services.  This should
+        ///     be called after any configuration changes.
+        /// </remarks>
+        /// <exception cref="Exception">
+        ///     When the engine is already started.
+        /// </exception>
+        public async Task StartAsync()
+        {
+            _dfsState.IsStarted = true;
+            if (_stopTasks.Count > 0)
+            {
+                throw new Exception("IPFS engine is already started.");
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['System.Exception' should not be thrown by user code.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)